Closes #263 #324: Add FxA service component and sample app

csadilek opened the pull request:
This brings in the prototype of the service-fxa module (based on the fxa client rust lib) and a sample app that uses it. The sample app shows a login button, and after successful login, displays some FxA profile data. This is pretty much the same as the iOS version hosted at https://github.com/mozilla/application-services/tree/master/sandvich-ios.

*This is just a prototype:* 🙃 There are many known issues and topics to discuss. I think we should bring this in so we can collaborate on it. I will file a bunch of issues for the known limitations and ToDos.

Issues filed so far: https://github.com/mozilla-mobile/android-components/labels/%3CFxA%3E

sputnikci commented on samples/fxa/src/main/java/org/mozilla/samples/fxa/MainActivity.kt:
```diff
@@ -0,0 +1,82 @@
+        val data = intent.dataString
+
+        if (Intent.ACTION_VIEW == action && data != null) {
+            val txtView:TextView = findViewById(R.id.txtView)
```
\[ktlint\] WARNING: \[colon\-spacing\] Missing spacing after ":" in column 25

pocmo commented on README.md:
```diff
@@ -149,6 +149,8 @@ _Generic low-level UI components for building apps._
 
 _Components and libraries to interact with backend services._
 
+* 🔴 [**FxA**](components/service/fxa/README.md) - A library for integrating with Firefox Accounts.
```
Even though the folks working on this know what FxA is: Can we name this more clearly? Maybe just FirefoxAccounts (service-firefox-accounts)?

pocmo commented on components/service/fxa/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# jniLibs
+*.so
+!libjnidispatch.so
```
Can we move this to the root .gitignore file? I think we should get rid of all the project ones.

pocmo commented on components/service/fxa/README.md:
```diff
@@ -0,0 +1,24 @@
+## Motivation
+
+The goal of this library is to provide a generic set of components to support a variety of telemetry use cases. It tries to not be opinionated about dependency injection frameworks or http clients. The only dependency is ``support-annotations`` to ensure code quality.
+
```
> The goal of this library is to provide a generic set of components to support a variety of telemetry use cases.

Is it? 🙃

pocmo commented on components/service/fxa/build.gradle:
```diff
@@ -0,0 +1,68 @@
+            artifact '/project.fxaclient.builds.[revision]/artifacts/public/bin/[organisation]/[module]_[revision].[ext]'
+        }
+    }
+    mavenCentral()
```
Do we need mavenCentral() here for something? jna maybe? Is that not on jcenter?

csadilek replied:
This shouldn't be needed. I will remove it.

pocmo commented on components/service/fxa/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,8 @@
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<manifest package="org.mozilla.fxa"
```
Following the pattern of our other libraries this should be: `mozilla.components.service.fxa` (Telemetry is a special case currently as it existed before)

pocmo commented on components/service/fxa/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# jniLibs
+*.so
+!libjnidispatch.so
```
libjnidispatch.so is something we commit to the repo?

csadilek replied:
Yeah, we can't get that easily from anywhere right now. I will file a follow-up to find a solution.

pocmo commented on samples/fxa/src/main/java/org/mozilla/samples/fxa/MainActivity.kt:
```diff
@@ -0,0 +1,82 @@
+        config?.let {
+            account = FirefoxAccount(it, CLIENT_ID)
+
+            val btn = findViewById<View>(R.id.button) as Button
```
nit: The method is generic to avoid the cast :) You may want:

```Kotlin
val btn = findViewById<Button>(R.id.button)
```

pocmo commented on components/service/fxa/src/main/java/mozilla/components/fxa/FirefoxAccount.kt:
```diff
@@ -0,0 +1,133 @@
+        } else {
+            Log.e("FirefoxAccount.init", e.consumeMessage())
+            this.rawPointer = null
+        }
```
All those checks make me wonder if we could turn Error into a sealed class with distinct types. With that we could just do something like:

```Kotlin
when (result) {
    is Success -> ... // Kotlin will do a nice smart cast here
    is ...
}
```

pocmo commented on components/service/fxa/src/main/java/mozilla/components/fxa/FirefoxAccount.kt:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    fun beginOAuthFlow(redirectURI: String, scopes: Array<String>, wantsKeys: Boolean): String? {
+        val scope = TextUtils.join(" ", scopes)
```
nit: TextUtils is not needed in Kotlin: [joinToString](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/join-to-string.html)

csadilek replied:
Fixed.

pocmo commented on components/service/fxa/src/main/java/mozilla/components/fxa/FirefoxAccount.kt:
```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    fun getTokenServerEndpointURL(): String? {
```
I wonder how painful it's going to be that everything and the world returns nullables.

pocmo commented on samples/fxa/src/main/java/org/mozilla/samples/fxa/MainActivity.kt:
```diff
@@ -0,0 +1,82 @@
+                account?.beginOAuthFlow(REDIRECT_URL, arrayOf("profile"), false)?.let {
+                    openAuthTab(it)
+                }
+            }
```
I agree with your issue about making this asynchronously. This synchronous chain of nullables is not fun :)

csadilek replied:
👍

pocmo commented:
Only the name... that's something I would like us to change now 🙃

csadilek commented on components/service/fxa/src/main/java/mozilla/components/fxa/Error.kt:
```diff
@@ -0,0 +1,58 @@
+     */
+    fun isSuccess(): Boolean {
+        return this.code == 0
+    }
```
https://github.com/mozilla-mobile/android-components/issues/349

csadilek commented on components/service/fxa/src/main/java/mozilla/components/fxa/Config.kt:
```diff
@@ -0,0 +1,38 @@
+            }
+        }
+
+        fun custom(content_base: String): Config? {
```
I'd say all of these things needs docs, yes. Let's get that API sorted next.

csadilek commented:
@pocmo I've addressed your issues, minus one. I did rename the packages, but kept the module names: service-fxa and samples-fxa. Mozilla uses fxa everywhere:

https://github.com/mozilla/fxa
https://github.com/mozilla/application-services/tree/master/fxa-rust-client
https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Firefox_Accounts/Introduction (Firefox Account (FxA))

The Firefox account libs all have fxa in the name. I think we should go with what the team prefers here. Do you feel strongly about this? I also slightly prefer the shorter name.

sputnikci commented on samples/fxa/src/main/java/org/mozilla/samples/fxa/MainActivity.kt:
```diff
@@ -0,0 +1,82 @@
+import mozilla.components.service.fxa.FirefoxAccount
+import android.support.customtabs.CustomTabsIntent
+import android.view.View
+import android.widget.Button
```
\[ktlint\] WARNING: \[no\-unused\-imports\] Unused import in column 1

pocmo commented:
> The Firefox account libs all have fxa in the name. I think we should go with what the team prefers here. Do you feel strongly about this? I also slightly prefer the shorter name.

I think that's fine and I wouldn't want to rename any of that - especially because it's internal. I'm mostly concerned about this _here_ being a public framework to be consumed by external partners and we are forcing everyone to do the mental mapping of FxA -> Firefox Accounts - something that (only externally) is not really clear. I definitely won't block you from merging it as-is because of that. 😃
